cpp/include/cugraph/graph_mask.hpp

```diff
@@ -0,0 +1,382 @@
+    : handle_(other.handle_),
+      n_vertices_(other.n_vertices_),
+      n_edges_(other.n_edges_),
+      edges_(other.edges_, other.handle_.get_stream()),
+      vertices_(other.vertices_, other.handle_.get_stream()),
+      complement_(other.complement_)
+  {
+  }
+
+  graph_mask_t& operator=(graph_mask_t&&) noexcept = default;
+  graph_mask_t& operator=(graph_mask_t const& other) = default;
+
+  /**
+   * Determines whether the 1 bit in a vertex or edge position
+   * represents an inclusive mask or exclusive mask. Default is
+   * an inclusive mask (e.g. 1 bit means the corresponding vertex
+   * or edge should be included in computations).
+   * @return
+   */
+  bool is_complemented() const { return complement_; }
+
+  /**
+   * Whether or not the current mask object has been initialized
+   * with an edge mask.
+   * @return
+   */
+  bool has_edge_mask() const { return get_edge_mask().has_value(); }
+
+  /**
+   * Whether or not the current mask object has been initialized
+   * with a vertex mask.
+   * @return
+   */
+  bool has_vertex_mask() const { return get_vertex_mask().has_value(); }
+
+  /**
+   * Returns the edge mask if it has been initialized on the instance
+   * @return
+   */
+  std::optional<mask_t const*> get_edge_mask() const
+  {
+    return edges_.size() > 0 ? std::make_optional<mask_t const*>(edges_.data()) : std::nullopt;
+  }
+
+  /**
+   * Retuns the vertex mask if it has been initialized on the instance
+   * @return
+   */
+  std::optional<mask_t const*> get_vertex_mask() const
+  {
+    return vertices_.size() > 0 ? std::make_optional<mask_t const*>(vertices_.data())
+                                : std::nullopt;
+  }
+
+  vertex_t get_n_vertices() { return n_vertices_; }
+
+  edge_t get_n_edges() { return n_edges_; }
+
+  edge_t get_edge_mask_size() const { return n_edges_ >> log_bits<mask_t>(); }
+
+  vertex_t get_vertex_mask_size() const { return n_vertices_ >> log_bits<mask_t>(); }
+
+  void initialize_edge_mask(bool init = 0)
+  {
+    if (!has_edge_mask()) {
+      allocate_edge_mask();
+      RAFT_CUDA_TRY(cudaMemsetAsync(edges_.data(),
+                                    edges_.size() * sizeof(mask_t),
+                                    std::numeric_limits<mask_t>::max() * init,
+                                    handle_.get_stream()));
+    }
+  }
+
+  void initialize_vertex_mask(bool init = 0)
+  {
+    if (!has_vertex_mask()) {
+      allocate_vertex_mask();
+      RAFT_CUDA_TRY(cudaMemsetAsync(vertices_.data(),
+                                    vertices_.size() * sizeof(mask_t),
+                                    std::numeric_limits<mask_t>::max() * init,
+                                    handle_.get_stream()));
+    }
+  }
+
+  /**
+   * Initializes an edge mask by allocating the device memory
+   */
+  void allocate_edge_mask()
+  {
+    if (edges_.size() == 0) {
+      edges_.resize(get_edge_mask_size(), handle_.get_stream());
+      clear_edge_mask();
```

Comment:
This is really just for completeness- if we initialize a mask to some value (it can be 0 or 1 and doesn't depend at all on whether it's a complement), it's nice to set the mask back to 0. That said, it's not even being called anywhere.